### DISCUSSION TITLE
Exposed function not connecting or not firing properly

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [cefsharp/CefSharp.Dom#55] [Q&A]
Trying to connect a click event on an `<a/>` element.

I am positive the `Context` is good, and I debugged in and know I have a valid anchor element instance.

The expose function is not failing for any reason, and apparently the element event connection is being made.

But I get no callback occur back into my code for whatever reason.

First, I gather I need to maintain the context instance; I am. And maintain that instance until the next Browser address change, at which point I dispose of the old one, which I presume is also like clearing the exposed functions cache, and generate a new context on the new page. I rinse and repeat the whole process, connecting the function, etc.

```csharp
private void OnLogoutClick()
{
    // TODO: respond to logout
}

protected virtual async Task<bool> TryVerifyLogoutAsync(ChromiumWebBrowser browser)
{
    var context = Context;
    Debug.Assert(context is not null, $"{nameof(context)} expected, was null.");

    if (!browser.Address.Equals(LoginUrl, InvariantCultureIgnoreCase))
    {
        var aLogout = await Context.QuerySelectorAsync<HtmlAnchorElement>(S.Logout);
        await Context.ExposeFunctionAsync("onLogoutClick", OnLogoutClick);
        await aLogout.AddEventListenerAsync("click", "onLogoutClick");
        return true;
    }

    return false;
}
```

I'm assuming it is `"click"` at least according to [_this documentation_](https://www.w3schools.com/jsref/dom_obj_event.asp), but maybe I am missing something there. Or perhaps I also need to maintain the anchor element instance once I have obtained that as well, which would be understandable, in addition to its context.

**Comment by mwpowellhtx** [comment]
What I am wondering is, do I need to arrange `CefSettings` or other _Chromium flags_ to allow for things such. I started more than less in this [_minimal example_](https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.Wpf/Program.PublishSingleFile.cs) just to get some things on their feet, and for the most part, it's fantastic, it works quite well. Except apparently in a couple of areas, maybe I need to configure a bit differently.

**Reply by mwpowellhtx** [reply]
Perhaps needing some degree of multithreading enabled and configured as well to allow for async callbacks? I forget exactly where I found that snippet, probably among the same minimal examples.

```csharp
var settings = new CefSettings();
settings.MultiThreadedMessageLoop = false; // This defaults to true

Cef.Initialize(settings);

// -For WPF use[DispatcherTimer](https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatchertimer?view=netframework-4.8)
// -For WinForms use[Timer](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.timer?view=netframework-4.8)

// Set the timer Interval to 30 times per second, can be increased to 60 if required
// For WPF
timer.Interval = TimeSpan.FromMilliseconds(1000 / 30);
// For WinForms
timer.Interval = 1000 / 30;
timer.Tick += UiThreadTimerTick;
timer.Start();

// Before closing your app
// Calling Cef.DoMessageLoopWork() after Cef.Shutdown has been called will result in an access violation, make sure you stop you timer first.
timer.Tick -= UiThreadTimerTick;
timer.Stop();
```

**Comment by amaitland** [comment]
> First, I gather I need to maintain the context instance; I am. And maintain that instance until the next Browser address change, at which point I dispose of the old one, which I presume is also like clearing the exposed functions cache, and generate a new context on the new page

The expected usage is a single `DevToolsContext` per browser instance. Use the same instance until the browser is disposed. `Functions` should persist even after navigations. 

The scenario you've described is not one that I've ever tested. This is not the standard use case for `Puppeteer`.

**Comment by mwpowellhtx** [comment]
I see, one instance per browser instance. So, ostensibly, I should rather position such a thing as a `UserControl.Resource`, let's say,, if that is where I am hosting the browser instance?

Yes, I have generally been keeping one context per function or method body, not as you described...
